drizzlepac/haputils/pipeline_graphics.py

```diff
@@ -0,0 +1,411 @@
+from bokeh.plotting import figure, output_file, save
+from bokeh.models import ColumnDataSource, Label, Range1d
+
+import numpy as np
+
+from stsci.tools import logutil
+
+from .pandas_utils import PandasDFReader
+from . import graph_utils
+from .graph_utils import HAPFigure
+
+
+
+MSG_DATEFMT = '%Y%j%H%M%S'
+SPLUNK_MSG_FORMAT = '%(asctime)s %(levelname)s src=%(name)s- %(message)s'
+log = logutil.create_logger(__name__, level=logutil.logging.NOTSET, stream=sys.stdout,
+                            format=SPLUNK_MSG_FORMAT, datefmt=MSG_DATEFMT)
+                            
+__taskname__ = 'pipeline_graphics'
+
+"""
+NOTES
+-----
+Code for generating vectors in Bokeh was found in:
+
+ https://github.com/holoviz/holoviews/blob/master/holoviews/plotting/bokeh/chart.py#L309
+
+This code shows how they define the arrow heads for the vector plots.  A
+'non-invasive' way to use the Bokeh Segment Glyph to create a vector plot would
+be to define the arrow heads as 2 additional segments starting at the endpoint
+of the vector. The new segments would then be added to the data points to be
+plotted by Segment to appear as an arrow head on each segment.
+
+
+
+# Auto-scaling computation for Quiver plot from matplotlib.
+# https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/quiver.py#L666-673
+ 
+# self.N : number of points to be plotted
+# amean : mean of vector lengths
+# self.span : width of plot
+
+            sn = max(10, math.sqrt(self.N))
+            if self.Umask is not ma.nomask:
+                amean = a[~self.Umask].mean()
+            else:
+                amean = a.mean()
+            # crude auto-scaling
+            # scale is typical arrow length as a multiple of the arrow width
+            scale = 1.8 * amean * sn / self.span
+
+Example from bokeh.org on how to create a tabbed set of plots:
+
+.. code:: python
+
+    from bokeh.io import output_file, show
+    from bokeh.models import Panel, Tabs
+    from bokeh.plotting import figure
+
+    output_file("slider.html")
+
+    p1 = figure(plot_width=300, plot_height=300)
+    p1.circle([1, 2, 3, 4, 5], [6, 7, 2, 4, 5], size=20, color="navy", alpha=0.5)
+    tab1 = Panel(child=p1, title="circle")
+
+    p2 = figure(plot_width=300, plot_height=300)
+    p2.line([1, 2, 3, 4, 5], [6, 7, 2, 4, 5], line_width=3, color="navy", alpha=0.5)
+    tab2 = Panel(child=p2, title="line")
+
+    tabs = Tabs(tabs=[ tab1, tab2 ])
+
+    show(tabs)
+
+
+"""
+
+# -------------------------------------------------------------------------------
+# Generate the Bokeh plot for the pipeline astrometric data.
+#
+HOVER_COLUMNS = ['gen_info.imgname',
+                 'gen_info.instrument',
+                 'gen_info.detector',
+                 'gen_info.filter',                
+                 'header.DATE-OBS',
+                 'header.RA_TARG',
+                 'header.DEC_TARG',
+                 'header.GYROMODE',
+                 'header.EXPTIME',
+                 'fit_results.aligned_to']
```

Comment:
Maybe I need to add more docs.  You get all the "gen_info" and "header.*" columns available in the Pandas DataFrame for free in pandas_utils, as well as the 'inst_det' and 'colormap' columns.  While this list can serve the same purpose and are closely related, this is slightly different than generating the hover tooltips.  I did not protect against dups in the hover tools tips list.

You get a few hover tooltips for free in

Reply:
This does not seem to hurt anything even though there will be duplicate columns in the output dataframe.  I can clean this up in the pandas_utils.py.  This is really just a matter of different style of different folks coding. This is not a blocker.

Reply:
Fixed in 057ba8a.